### DISCUSSION TITLE
Map PSQL enum with Kotlin enum

### DISCUSSION THREAD

**Discussion by Distractic** [komapper/komapper#1046] [Q&A] [answered]
Hello !

I'm trying to map a column with an enum property.
I'm using PSQL database.

Kotlin classes: 
```kt
public enum class Rank {
    PLAYER,
    ADMIN
}

@KomapperEntity
@KomapperTable("player")
public data class Player(
    @KomapperId
    val uuid: UUID,
    val rank: Rank,
)
```

SQL file:
```sql
CREATE TYPE rank AS ENUM ('PLAYER', 'ADMIN');

CREATE TABLE player
(
    uuid uuid PRIMARY KEY,
    rank rank NOT NULL
);
```

Gradle KSP arguments:
```gradle
ksp {
    arg("komapper.enumStrategy", "name")
}
```

However, I have this error when I'm trying to insert:
```
io.r2dbc.postgresql.ExceptionFactory$PostgresqlBadGrammarException: [42804] column "rank" is of type rank but expression is of type character varying
```

So how can I map my Kotlin enum with my PSQL enum ?

Thanks in advance

**Comment by nakamura-to** [comment]
The database enum type is currently not supported.

However, we would like to consider supporting that in the future.
For now, we plan to indicate a mapping to the database enum type by writing something like `@KomapperEnum(EnumType.ENUM)`.

```kotlin
@KomapperEntity
@KomapperTable("player")
public data class Player(
    @KomapperId
    val uuid: UUID,
    @KomapperEnum(EnumType.ENUM)
    val rank: Rank,
)
```

**Reply by Distractic** [reply]
I don't have the enum type `ENUM` for `EnumType`. I only have `NAME`, `ORDINAL` and `PROPERTY`

**Reply by nakamura-to** [reply]
The enum type `ENUM` for `EnumType` is an element to be introduced in the future.

**Reply by Distractic** [reply]
In this case I will wait. Thanks

**Reply by nakamura-to** [reply]
Your request is supported by https://github.com/komapper/komapper/pull/1079

**Reply by Distractic** [reply]
I'm trying to use the enum but I have an error about the type

```kt
public enum class Rank {
    PLAYER,
    ADMIN
}

public class RankType : R2dbcUserDefinedDataType<Rank> {

    override val name: String = "rank"

    override val klass: KClass<Rank> = Rank::class

    override val r2dbcType: Class<*> = String::class.javaObjectType

    override fun getValue(row: Row, index: Int): Rank? {
        return row[index, r2dbcType]?.let {
            Rank.entries.find { rank -> rank.name == it }
        }
    }

    override fun getValue(row: Row, columnLabel: String): Rank? {
        return row[columnLabel, r2dbcType]?.let {
            Rank.entries.find { rank -> rank.name == it }
        }
    }

    override fun toString(value: Rank): String {
        return value.name
    }

    override fun setValue(statement: Statement, name: String, value: Rank) {
        statement.bind(name, value.name)
    }

    override fun setValue(statement: Statement, index: Int, value: Rank) {
        statement.bind(index, value.name)
    }
}
```

I have the following error: 
```kt
io.r2dbc.postgresql.ExceptionFactory$PostgresqlBadGrammarException: [42804] column "rank" is of type rank but expression is of type character varying
```

For the SQL table:
```sql
CREATE TYPE rank AS ENUM ('PLAYER', 'ADMIN');

-- Create table to store players
CREATE TABLE player
(
    uuid     uuid PRIMARY KEY,
    rank     rank NOT NULL,
    language varchar(50) NOT NULL
);
```

I think I miss something with the properties of user data type class
Can you help me to know what is the problem please ?

**Reply by nakamura-to** [reply]
You define the `r2dbcType` property as a String class:

```kotlin
override val r2dbcType: Class<*> = String::class.javaObjectType
```

But, you should it as a rank class:

```kotlin
override val r2dbcType: Class<*> = rank::class.javaObjectType
```

Then don't forget to use EnumCodec!
https://github.com/pgjdbc/r2dbc-postgresql#postgres-enum-types

In addition, `@KomapperEnum(EnumType.Type)` must be annotated to the `rank` property of your entity class as follows:

```kotlin
@KomapperEnum(EnumType.Type)
val rank;
```
See also https://www.komapper.org/docs/reference/entity-class/#komapperenum

**Reply by Distractic** [reply]
Ok thanks
Now I don't understand  how to mix the codec configuration with the R2DBC constructor

For my test, I'm using this:
```kt
@BeforeTest
fun onBefore() = runBlocking {
    database = R2dbcDatabase(createConnectionOptions(psqlContainer))
}

fun createConnectionOptions(container: PostgreSQLContainer<*>): ConnectionFactoryOptions {
    return ConnectionFactoryOptions.builder()
        .option(ConnectionFactoryOptions.DRIVER, PostgreSqlDialect.driver)
        .option(ConnectionFactoryOptions.USER, container.username)
        .option(ConnectionFactoryOptions.PASSWORD, container.password)
        .option(ConnectionFactoryOptions.HOST, container.host)
        .option(ConnectionFactoryOptions.PORT, container.getMappedPort(PostgreSQLContainer.POSTGRESQL_PORT))
        .option(ConnectionFactoryOptions.DATABASE, container.databaseName)
        .build()
}
```

But the `R2dbcDatabase` can't receive a `PostgresqlConnectionFactory` in the constructor.
Can you tell me how can I pass the PSQL connection factory with codec to the R2DBC client ?

**Reply by nakamura-to** [reply, marked as answer]
`R2dbcDatabase` can receive a `PostgresqlConnectionFactory`.
Try the following code:

```kotlin
    @BeforeTest
    fun onBefore() = runBlocking {
        database = R2dbcDatabase(
            connectionFactory = createPostgresqlConnectionFactory(psqlContainer), 
            dialect = PostgreSqlR2dbcDialect()
        )
    }

    fun createPostgresqlConnectionFactory(container: PostgreSQLContainer<*>): PostgresqlConnectionFactory {
        val options = createConnectionOptions(container)
        val configuration =
            PostgresqlConnectionFactoryProvider.builder(options)
                .codecRegistrar(EnumCodec.builder().withEnum("rank", Rank::class.java).build())
                .build()
        return PostgresqlConnectionFactory(configuration)
    }

    fun createConnectionOptions(container: PostgreSQLContainer<*>): ConnectionFactoryOptions {
        return ConnectionFactoryOptions.builder()
            .option(ConnectionFactoryOptions.DRIVER, PostgreSqlDialect.driver)
            .option(ConnectionFactoryOptions.USER, container.username)
            .option(ConnectionFactoryOptions.PASSWORD, container.password)
            .option(ConnectionFactoryOptions.HOST, container.host)
            .option(ConnectionFactoryOptions.PORT, container.getMappedPort(PostgreSQLContainer.POSTGRESQL_PORT))
            .option(ConnectionFactoryOptions.DATABASE, container.databaseName)
            .build()
    }
```

**Reply by Distractic** [reply]
Thanks, but I always have the error `column "rank" is of type rank but expression is of type character varying`, so I don't understand the problem.. This is the [commit](https://github.com/Rushyverse/core/commit/5184dfee3e10d2e51afe9e476ded3c51df0a868a), when you have time, can you take a look, please?

**Reply by nakamura-to** [reply]
https://github.com/Rushyverse/core/commit/5184dfee3e10d2e51afe9e476ded3c51df0a868a#diff-1213fca9f58bd2fbf91c6775e2c5998ba238f28fd741dbe4d0b884052832f5f9R39
https://github.com/Rushyverse/core/commit/5184dfee3e10d2e51afe9e476ded3c51df0a868a#diff-1213fca9f58bd2fbf91c6775e2c5998ba238f28fd741dbe4d0b884052832f5f9R43

They should be `statement.bind(name, value)` or `statement.bind(index, value)`.

**Reply by Distractic** [reply]
Ah yes, that's it!

Your [documentation ](https://www.komapper.org/docs/reference/data-type/#user-defined-data-types-for-r2dbc)is for a data class,  not for an Enum class, so I foolishly followed the example with the `Age(int)`.
That could be beneficial to have an example with an Enum too.

Thanks a lot!
